client/lib/cart/store/cart-synchronizer.js

```diff
@@ -52,11 +52,26 @@ function preprocessCartForServer( cart ) {
 		'products',
 		'coupon',
 		'is_coupon_applied',
+		'is_coupon_removed',
 		'currency',
 		'temporary',
 		'extra'
 	);
 
+	if (
+		! newCart.coupon &&
+		! newCart.is_coupon_applied &&
+		! newCart.is_coupon_removed &&
+		typeof document !== 'undefined'
+	) {
+		const coupon = url.parse( document.URL, true ).query.coupon;
+
+		if ( coupon ) {
+			newCart.coupon = coupon;
```

Comment:
is this assignment not redundant?

Reply:
@dmsnell `newCart.coupon` is set if there isn't one but we got coupon from URL, makes sense to me. What doesn't feel right here is that we are checking the URL on every cart-related operation. What would be better is  kind of a redux middleware that would check the URL and explicitly dispatch `applyCoupon` and remove the `coupon` bit from the URL after it does it's job.

Reply:
haha. I unintentionally proved my point about mutation. I literally did not read this comment until writing my response below but I _did_ end up confused about the split assignment because I didn't see the second `coupon` assignment…